test/transaction_tracker.go

```diff
@@ -0,0 +1,193 @@
+func (tr *TransactionTracker) trackTransactions() error {
+	for {
+		select {
+		case h := <-tr.heads:
+			b, err := tr.client.BlockByHash(context.Background(), h.Hash())
+			if err != nil {
+				return err
+			}
+			// If we have transactions then process them
+			if len(b.Transactions()) > 0 {
+				tr.newTxs.L.Lock()
+				for _, t := range b.Transactions() {
+					tr.processed[t.Hash()] = b
+				}
+				// signal
+				tr.newTxs.Signal()
+				tr.newTxs.L.Unlock()
+			}
+		case err := <-tr.sub.Err():
+			// Will be nil if closed by calling Unsubscribe()
+			return err
+		case <-tr.stopCh:
+			return nil
+		}
+	}
+}
+
+// AwaitTransactions waits for the transactions listed in hashes to be
+// processed, it will return the ctx.Err() if ctx expires before all the
+// transactions in hashes were processed or ErrStopped if StopTracking is
+// called before all the transactions in hashes were processed.
+func (tr *TransactionTracker) AwaitTransactions(ctx context.Context, hashes []common.Hash) error {
+	hashmap := make(map[common.Hash]struct{}, len(hashes))
+	for i := range hashes {
+		hashmap[hashes[i]] = struct{}{}
+	}
+	tr.newTxs.L.Lock()
+	defer tr.newTxs.L.Unlock()
+	txsFound := make(chan struct{})
+	var exitErr error
+	// This go-routine will either call Signal when the context expires or the
+	// tracker is closed, allowing us to wake from waiting on these events, or
+	// simply exit if the awaited transactions are processed before the context
+	// expires or the tracker is closed.
+	tr.wg.Add(1)
+	go func() {
+		defer tr.wg.Done()
+		// Note that we have already locked tr.newTxs.L in the creating
+		// go-routine and it will only be unlocked when Wait is called from the
+		// creating go-routine, this means Wait must have been called before we
+		// are able to send our signal.
+		select {
+		case <-ctx.Done():
+			tr.newTxs.L.Lock()
+			defer tr.newTxs.L.Unlock()
+			exitErr = ctx.Err()
+			tr.newTxs.Signal()
+		case <-tr.stopCh:
+			tr.newTxs.L.Lock()
+			defer tr.newTxs.L.Unlock()
+			exitErr = errStopped
+			tr.newTxs.Signal()
+		case <-txsFound:
+			return
+		}
+	}()
+
+	for {
+		// Delete processed transactions from hashmap
+		for hash := range hashmap {
+			_, ok := tr.processed[hash]
+			if ok {
+				delete(hashmap, hash)
+			}
+		}
+		// If there are no transactions left then they have all been processed.
+		if len(hashmap) == 0 {
+			close(txsFound) // Signal to the go routine that it can exit
+			return nil
+		}
+		tr.newTxs.Wait()
+		// Check the exit error to see if the context expired.
+		if exitErr != nil {
+			return exitErr
+		}
+	}
+}
+
+// StopTracking shuts down all the goroutines in the tracker.
+func (tr *TransactionTracker) StopTracking() error {
+	if tr.sub == nil {
+		return errors.New("attempted to stop already stopped tracker")
+	}
+	tr.sub.Unsubscribe()
+	close(tr.stopCh)
+	tr.wg.Wait()
+	tr.wg = sync.WaitGroup{}
+
+	// Free fields
+	tr.client = nil
```

Comment:
Shouldn't these be cleaned up by the GC once the TransactionTracker can no longer be reached

Reply:
Those fields are set in StartTracking so if we were to start and stop a tracker those fields would remain uncollected until start was called again. However the freeing of them definitely seems an unnecessary optimisation (if it can even be called that), I can't remember what I was thinking when I wrote that code. So I think we can safely just remove these lines.